test/integration/retry_budget_integration_test.cc

```diff
@@ -0,0 +1,284 @@
+#include "envoy/config/bootstrap/v3/bootstrap.pb.h"
+#include "envoy/type/http.pb.h"
+
+#include "test/integration/http_protocol_integration.h"
+
+#include "fake_upstream.h"
+#include "http_integration.h"
+
+namespace Envoy {
+namespace {
+
+class RetryBudgetIntegrationTest : public HttpProtocolIntegrationTest {
```

Comment:
This integration test setup is particularly finnicky since we have to have multiple connections/streams/codecs at once, and the integration test common code isn't really built for it. Mostly complicated for HTTP1 upstreams in terms of guessing which connection will be used.

Reply:
I wanted an integration test because the retry budgets seem to be very intricately related to two complicated parts of upstream behavior:
1. Upstream codec client behavior
2. Connection pool tracking (i.e. pending/active request tracking)